### PR TITLE
[AMD][Quantization] Add `int4fp8_moe` online quantization on ROCm

### PR DESCRIPTION
**Pull request by fxmarty-amd**
As per title, this PR supersedes https://github.com/sgl-project/sglang/pull/6238.

This PR implements loading MOE models checkpoints in high-precision (fp16, bf16), quantizing online the MOE experts to int4 and the attention projections to float8.

During inference the int4 moe weights are upcasted to float8 in order to use fp8 math.

Runnable with
```bash
export MODEL_ID="/large_models/mistralai_Mixtral-8x7B-Instruct-v0.1"
python3 -m sglang.launch_server --model-path ${MODEL_ID} --tensor-parallel-size 1 --disable-cuda-graph --quantization quark_int4fp8_moe
```

Left to do:
- [x] better tests
- [x] shard non-quantized weights before online quantization
- [x] eval grok1
- [x] benchmark latency/throughput
- [x] doc
- [x] merge main
- [ ] Add grok-1 test

### CODE REVIEW

File: python/sglang/srt/layers/quark_utils.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+def quantize_fp8_scale_tensorwise(w: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    FP8_MAX = 448.0
+    scale = w.abs().amax().float() / FP8_MAX
+    scaled = (w / scale).clamp(-FP8_MAX, FP8_MAX).to(torch.float8_e4m3fn)
+    return scaled, scale
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quark_utils.py:17]: irrelevant

File: python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.float16, torch.bfloat16]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 70
+
+    @classmethod
+    def get_name(self) -> str:
+        return "w4a8_int4fp8"
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py]: fixed in 3df5e1ec241728f3e2397a2e7ee27e57c7ca9cb2

File: python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+    def from_config(cls, config: Dict[str, Any]) -> "QuarkInt4Fp8Config":
+        quant_method = cls.get_from_keys(config, ["quant_method"])
+        activation_scheme = cls.get_from_keys(config, ["activation_scheme"])
+        ignored_layers = cls.get_from_keys_or(config, ["ignored_layers"], None)
+        weight_block_size = cls.get_from_keys_or(config, ["weight_block_size"], None)
+        return cls(
+            activation_scheme=activation_scheme,
+            ignored_layers=ignored_layers,
+            weight_block_size=weight_block_size,
+        )
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py]: outdated

File: python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+            activation=activation,
+            apply_router_weight_on_input=apply_router_weight_on_input,
+            use_fp8_w8a8=True,
+            w1_scale=(
+                layer.w13_fp8_scale_inv
+                if self.block_quant
+                else layer.w13_fp8_scale
+            ),
+            w2_scale=(
+                layer.w2_fp8_scale_inv if self.block_quant else layer.w2_fp8_scale
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py]: fixed in a4757098793501156b68f6f48c5a6582f2365e94

File: python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+            loaded_weight: torch.Tensor,
+            weight_name: str,
+            shard_id: str,
+            expert_id: int,
+        ):
+            fp8_w, fp8_scale = quantize_fp8_scale_tensorwise(loaded_weight)
+            int4_w, int4_scale = quantize_int4_scale_columnwise(loaded_weight)
+
+            int4_w = pack_int4_to_int32(int4_w)
+            int4_scale /= fp8_scale
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py]: irrelevant

File: python/sglang/srt/layers/quark_utils.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def quantize_int4_scale_columnwise(w: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    S4_MAX = 7
+    w_flat = w.reshape(-1, w.shape[-1]).float()
+    scale = w_flat.abs().amax(axis=-1) / S4_MAX
+    scaled = torch.round(w_flat / scale[:, None]).to(torch.int8).clamp(-S4_MAX, S4_MAX)
+    return scaled.reshape(w.shape), scale.reshape(w.shape[:-1])
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quark_utils.py]: irrelevant

File: python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+            layer.w2_weight,
+            topk_weights=topk_weights,
+            topk_ids=topk_ids,
+            inplace=inplace and not no_combine,
+            activation=activation,
+            apply_router_weight_on_input=apply_router_weight_on_input,
+            use_fp8_w8a8=True,
+            w1_scale=(
+                layer.w13_fp8_scale_inv
+                if self.block_quant
```
[the remark this answers is not in the document]
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/quark_w4a8_int4fp8.py]: fixed in a4757098793501156b68f6f48c5a6582f2365e94

File: python/sglang/srt/layers/quark_utils.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+"""
```
Comment (HaiShaw) [inline python/sglang/srt/layers/quark_utils.py:1]: rename this file to `python/sglang/srt/layers/quantization/int4fp8_utils.py`
nothing quark specific

**Review by HaiShaw — CHANGES_REQUESTED** [review]
`int4fp8` for Linear layer shall be out of the scope.
`projections in the attention layers have their weights quantized online to float8 directly`, shall also be out of the scope.

Please also provide the accuracy data on mixtral (compare with mixtral bfloat16 baseline, mixtral with --quantization fp8), grok1 (compare with bfloat16 base line, grok1 with --quantization fp8)

File: python/sglang/srt/layers/quantization/int4fp8_moe.py [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+            weight_scale=layer.weight_scale,
+            input_scale=layer.input_scale,
+            bias=bias,
+            cutlass_fp8_supported=False,
+            use_per_token_if_dynamic=False
+        )
+
+
+class QuarkInt4Fp8MoEMethod:
+    """MoE method for INT8.
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/int4fp8_moe.py]: int4?

File: python/sglang/srt/layers/quantization/int4fp8_moe.py [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+        elif isinstance(layer, FusedMoE):
+            return QuarkInt4Fp8MoEMethod(self)
+        
+        return None
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class QuarkInt4Fp8LinearMethod(LinearMethodBase):
```
Comment (BowenBao) [inline python/sglang/srt/layers/quantization/int4fp8_moe.py]: IIUC the `Int4Fp8` in name only refers to the overall quant method name, For linear layers this just performs fp8 quantization and no int4 right? Let's put a note in the docstring if that's the case.

**Comment by BowenBao** [comment]
> `int4fp8` for Linear layer shall be out of the scope. `projections in the attention layers have their weights quantized online to float8 directly`, shall also be out of the scope.
> 
> Please also provide the accuracy data on mixtral (compare with mixtral bfloat16 baseline, mixtral with --quantization fp8), grok1 (compare with bfloat16 base line, grok1 with --quantization fp8)

@HaiShaw I believe the linear layers are dynamically quantized to fp8, not int4. That includes 'projections in the attention layers', i.e. qkvo proj layers. In short this is the same spec for https://huggingface.co/amd/grok-1-W4A8KV8, with quantization done dynamically during original half precision ckpt loading.

We will apply your comments on naming suggestions and post the results on mixtral and grok1. If there are no more issues, please help land this PR.

**Comment by fxmarty-amd** [comment]
@HaiShaw @BowenBao 

Here are eval numbers from:

```bash
export MODEL_ID="/large_models_2/lmzheng_grok-1"
python -m sglang.launch_server --model-path ${MODEL_ID} --port 30000 --tensor-parallel-size 8
# python -m sglang.launch_server --model-path ${MODEL_ID} --port 30000 --tensor-parallel-size 8 --quantization fp8
# python -m sglang.launch_server --model-path ${MODEL_ID} --port 30000 --tensor-parallel-size 8 --quantization int4fp8_moe
```
&
```bash
cd benchmark/gsm8k && python3 bench_sglang.py --num-questions 2000 --parallel 2000
```

```
--quantization int4fp8_moe

Accuracy: 0.804
Invalid: 0.000
Latency: 27.976 s
Output throughput: 6957.681 token/s

original

Accuracy: 0.829
Invalid: 0.000
Latency: 41.939 s
Output throughput: 4644.659 token/s

--quantization fp8

Accuracy: 0.817
Invalid: 0.001
Latency: 30.232 s
Output throughput: 6463.059 token/s
```

File: python/sglang/srt/layers/quantization/__init__.py [hunk 11]
```diff
@@ -16,6 +16,7 @@ def override_quantization_method(self, *args, **kwargs):
 
 CompressedTensorsConfig = DummyConfig
 
+from python.sglang.srt.layers.quantization.quark_int4fp8_moe import QuarkInt4Fp8Config
```
Comment (yctseng0211) [inline python/sglang/srt/layers/quantization/__init__.py]: Could you explain the rationale for using
```from python.sglang.srt.layers.quantization...```
here instead of 
```from sglang.srt.layers.quantization...?```

**Comment by HaiShaw** [comment]
@fxmarty-amd

File: python/sglang/srt/layers/quantization/__init__.py (hunk 11 above)
Comment (fxmarty-amd) [reply python/sglang/srt/layers/quantization/__init__.py]: It is a typo, good spot

**Review by HaiShaw — APPROVED** [review]
